### ISSUE TITLE
No error tracking after response has been sent?

### ISSUE THREAD

**Issue by schnittchen** [appsignal/appsignal-elixir#285]
I'm seeing a number of errors `(Plug.Conn.AlreadySentError) the response was already sent` in the logs that are not showing up in the AppSignal app.

I suspect that's because the request error tracking is tied in some way to the `conn` "lifecycle"...

**Comment by schnittchen** [comment]
This is easily reproduced by adding a plug using send_resp/3 but _not_ halt/1 in a controller.

**Comment by jeffkreeftmeijer** [comment]
Hey @schnittchen,

Thanks for letting us know about this. It sounds like the transaction is closed before the request is completed, causing the error to be missed. Do you get a performance sample for this request in AppSignal? We [wrap](https://github.com/appsignal/appsignal-elixir/blob/develop/lib/appsignal/plug.ex#L14-L29) the whole endpoint, so we should be able to catch these exceptions.

We'll try to reproduce this and we'll get back to you as soon as we know more. To speed things up (if you have time); could you paste a minimal example controller that shows the problem?

Thanks!

**Comment by jeffkreeftmeijer** [comment]
I looked at the issue and found the easiest way to reproduce this is to raise after a `render/2`, which won't show up in AppSignal right now. 

```elixir
defmodule AppsignalPhoenixExampleWeb.PageController do
  use AppsignalPhoenixExampleWeb, :controller

  def index(conn, _params) do
    render conn, "index.html"
    raise "Error raised after render!"
  end
end
```

Although we wrap the whole endpoint, we don't catch these exceptions because we rely on `Plug.ErrorHandler.__catch__/2`, which is built to show error pages. To properly do that, it [doesn't try](https://github.com/elixir-plug/plug/blob/master/lib/plug/error_handler.ex#L86-L88) to render the error view when `@already_sent` is set. This makes it a bad match for us, because exceptions that happen after rendering should be reported to AppSignal.

To fix this, we'll stop relying on `__catch__/2`, and we'll keep this issue updated. Thanks again for letting us know about this, @schnittchen! <3

**Comment by jeffkreeftmeijer** [comment]
Hey @schnittchen,

We've released version 1.4.6 yesterday, which includes the patch for this issue. Could you upgrade by running `mix deps.update appsignal`, and let us know if exceptions raised after the response has been sent are caught properly by AppSignal? Thanks again for reporting this issue! :)

**Comment by schnittchen** [comment]
Tiny heads-up:

It looks like `mix deps.unlock appsignal; mix deps.get` or, shorter, `mix do: deps.unlock appsignal, deps.get` is a safer upgrade -- I had fewer changes in my mix.lock.
